pkg/cluster/internal/create/actions/config/config.go

```diff
@@ -62,39 +65,66 @@ func (a *Action) Execute(ctx *actions.ActionContext) error {
 		return errors.Wrap(err, "failed to get kubernetes version from node")
 	}
 
-	// get the control plane endpoint, in case the cluster has an external load balancer in
-	// front of the control-plane nodes
+	// get the control plane endpoint
 	controlPlaneEndpoint, err := nodes.GetControlPlaneEndpoint(allNodes)
 	if err != nil {
 		// TODO(bentheelder): logging here
 		return err
 	}
 
-	// get kubeadm config content
-	kubeadmConfig, err := getKubeadmConfig(
-		ctx.Config,
-		kubeadm.ConfigData{
-			ClusterName:          ctx.ClusterContext.Name(),
-			KubernetesVersion:    kubeVersion,
-			ControlPlaneEndpoint: controlPlaneEndpoint,
-			APIBindPort:          kubeadm.APIServerPort,
-			APIServerAddress:     ctx.Config.Networking.APIServerAddress,
-			Token:                kubeadm.Token,
-			PodSubnet:            ctx.Config.Networking.PodSubnet,
-		},
-	)
+	// create kubeadm init config
+	fns := []func() error{}
+
+	kubeadmClusterConfig := kubeadm.ConfigData{
+		ClusterName:          ctx.ClusterContext.Name(),
+		KubernetesVersion:    kubeVersion,
+		ControlPlaneEndpoint: controlPlaneEndpoint,
+		APIBindPort:          kubeadm.APIServerPort,
+		APIServerAddress:     ctx.Config.Networking.APIServerAddress,
+		Token:                kubeadm.Token,
+		PodSubnet:            ctx.Config.Networking.PodSubnet,
+		ControlPlane:         true,
+	}
+
+	fns = append(fns, func() error {
+		return writeKubeadmConfig(ctx.Config, kubeadmClusterConfig, node)
+	})
 
+	// create the kubeadm join configuration for secondary control plane nodes if any
+	secondaryControlPlanes, err := nodes.SecondaryControlPlaneNodes(allNodes)
 	if err != nil {
-		// TODO(bentheelder): logging here
-		return errors.Wrap(err, "failed to generate kubeadm config content")
+		return err
+	}
+	if len(secondaryControlPlanes) > 0 {
+		kubeadmClusterConfig.ControlPlane = true
+		// create the workers concurrently
+		for _, node := range secondaryControlPlanes {
+			node := node // capture loop variable
+			fns = append(fns, func() error {
+				return writeKubeadmConfig(ctx.Config, kubeadmClusterConfig, &node)
+			})
+		}
 	}
 
-	log.Debug("Using kubeadm config:\n" + kubeadmConfig)
+	// then create the kubeadm join config for the worker nodes if any
+	workers, err := nodes.SelectNodesByRole(allNodes, constants.WorkerNodeRoleValue)
+	if err != nil {
+		return err
+	}
+	if len(workers) > 0 {
+		kubeadmClusterConfig.ControlPlane = false
```

Comment:
this can cause all config templates to be controlplane=false (or rather there's a race)

that's because while the function calls will get a pass by value of this struct, they copy from the same shared instance of the struct and we assign to it here, and the goroutine scheduling is racy.

I'm fixing this in the patch to fix v1.11